tests/integration/src/compiler_test.rs

```diff
@@ -81,196 +50,517 @@ impl HirArtifact {
-            ..Default::default()
+            target_dir: None,
+            name: name.into(),
+            target: "wasm32-unknown-unknown".into(),
+            output_name: None,
+            source_code: source_code.into(),
         }
     }
+}
 
-    /// Set the Rust source code to compile a library Cargo project to Wasm module
-    pub fn rust_source_cargo_lib(
-        cargo_project_folder: PathBuf,
-        artifact_name: &str,
-        is_build_std: bool,
-        entry_func_name: Option<String>,
-    ) -> Self {
-        let expected_wasm_artifact_path = wasm_artifact_path(&cargo_project_folder, artifact_name);
-        // dbg!(&wasm_artifact_path);
-        let wasm_artifact_path = if !skip_rust_compilation(&cargo_project_folder, artifact_name)
-            || !expected_wasm_artifact_path.exists()
-        {
-            let manifest_path = cargo_project_folder.join("Cargo.toml");
-            let mut cargo_build_cmd = Command::new("cargo");
-            let compiler_workspace_dir = get_workspace_dir();
+/// The various types of input artifacts that can be used to drive compiler tests
+pub enum CompilerTestInputType {
+    /// A project that uses `cargo component build` to produce a Wasm module to use as input
+    CargoComponent(CargoTest),
+    /// A project that uses `cargo build` to produce a core Wasm module to use as input
+    Cargo(CargoTest),
+    /// A project that uses `rustc` to produce a core Wasm module to use as input
+    Rustc(RustcTest),
+}
+impl From<CargoTest> for CompilerTestInputType {
+    fn from(config: CargoTest) -> Self {
+        Self::Cargo(config)
+    }
+}
+impl From<RustcTest> for CompilerTestInputType {
+    fn from(config: RustcTest) -> Self {
+        Self::Rustc(config)
+    }
+}
+
+/// [CompilerTestBuilder] is used to obtain a [CompilerTest], and subsequently run that test.
+///
+/// Testing the compiler involves orchestrating a number of complex components. First, we must
+/// obtain the input we wish to feed into `midenc` for the test. Typically, we have some Rust
+/// source code, or a Cargo project, and we must compile that first, in order to get the Wasm
+/// module/component which will be passed to `midenc`. This first phase requires some configuration,
+/// and that configuration affects later phases (such as the name of the artifact produced).
+///
+/// Secondly, we need to prepare the [midenc_session::Session] object for the compiler. This is
+/// where we specify inputs, and various bits of configuration that are important to the test, or
+/// which are needed in order to obtain useful diagnostic output. This phase requires us to
+/// construct the base configuration here, but make it possible to extend/alter in each specific
+/// test.
+///
+/// Lastly, we must run the test, and in order to do this, we must know where our inputs and outputs
+/// are, so that we can fetch files/data/etc. as needed; know the names of things to be called, and
+/// more.
+pub struct CompilerTestBuilder {
```

Comment:
I love the new design! `CompilerTest` started small but quickly grew out of control. I was not happy with it for a long time but did not find a time to properly re-design it. Thank you!

Reply:
For how quickly and haphazardly we had to grow the test suite, I think your initial implementation has held up pretty well! It's made it through several changes/refactorings without too much pain in the process, which is all you really can ask for. The nice thing is that by the time I got around to the changes in this PR, the existing code made it very clear what the patterns were, so we have you to thank for doing the hard work of finding them first 😃